### DISCUSSION TITLE
setBackgroundMessageHandler behaviour appears inconsistent with docs

### DISCUSSION THREAD

**Discussion by jamie-unmind** [invertase/react-native-firebase#4946] [Q&A]
Hi there,

We are using `@react-native-firebase/messaging` 10.8.0 to send Notification+Data messages for iOS and Android.

According to the docs [here](https://rnfirebase.io/messaging/usage#message-handlers), `setBackgroundMessageHandler` should be called when the application is in a Background or Quit state. However, we are experiencing some behaviour that is inconsistent with this.

**When the application is backgrounded:**

- For iOS, we have to set `contentAvailable: true` in the `apns` section of the message in order to trigger `setBackgroundMessageHandler. The docs [seems to suggest](https://rnfirebase.io/messaging/usage#message-handlers) this should only be the case for Data only messages?
```javascript
{
      notification: {...},
      data: {...},
      apns: {        
        payload: {
          aps: {
            contentAvailable: true
          }
        }
      }
}
```

**When the application is quit (ie: user swipes it away):**

- For Android, we have to set `priority: "high"` in the `android` section of the message in order to trigger `setBackgroundMessageHandler. The docs [seems to suggest](https://rnfirebase.io/messaging/usage#message-handlers) this should only be the case for Data only messages?
```javascript
{
    notification: {...},
    data: {...},
    android: {
        priority: "high"
    }	
}
```
- For iOS, `setBackgroundMessageHandler` appears to never be triggered. This [issue comment](https://github.com/invertase/react-native-firebase/issues/3530#issuecomment-695367015) seems to suggest that once an iOS app is swiped away, `setBackgroundMessageHandler` cannot be triggered. However, [this matrix](https://rnfirebase.io/messaging/usage#message-handlers) in the docs suggests that `setBackgroundMessageHandler` **can** be called in a _Quit_ state, where Quit is defined as _"When the device is locked or application is not active or running. The user can quit an app by "swiping it away" via the app switcher UI on the device."_ (source [here](https://rnfirebase.io/messaging/usage#receiving-messages)). Is it possible to clarify which is the expected behaviour?

 Let me know if you need any more background info.

**Comment by mikehardy** [comment]
The FCM payload you are sending appears to be mixed (contains both notification and data chunks)

I believe for mixed payloads, the notification chunk is the important part, and it means that the firebase-ios-sdk / firebase-android-sdk will handle it and should post a device-local notification without ever giving control to your code - no handlers called

If/when the user interacts with the device-local notification in a way that opens your app, I believe that you will see it as an initial notification?

**Reply by jamie-unmind** [reply]
Yes our payloads are mixed.

Yes we are able to see when a user interacts with our notification in a way that opens the app - we have setup `getInitialNotification()` and `onNotificationOpenedApp()` in order to track when a user has tapped to open our notification.

However, we would also like to track when a user receives a notification so that we can compare no of notifications received against no of notifications interacted with. 

In this way, we want to be able to use `setBackgroundMessageHandler` to separately track when a notification is received regardless of whether the user has interacted with it. Should this be possible for mixed payload messages to iOS & Android, both when the application is backgrounded or quit (ie: swiped away)?

**Reply by mikehardy** [reply]
Unless I am mistaken mixed payloads do not trigger handlers on delivery. They only pop a notification center local notification. I think our docs - the table you linked - is not correct unfortunately :scream: 

https://firebase.google.com/docs/cloud-messaging/concept-options#notification-messages-with-optional-data-payload

**Reply by jamie-unmind** [reply]
Ah ok, does this mean we'd be reliant on our native code to track when a mixed payload message is received?

Is it unadvisable to use `contentAvailable: true` and `priority: "high"` to try to force our background message handler to trigger with a mixed payload message?

**Reply by mikehardy** [reply]
I'm not sure if the native SDKs give you any hooks in mixed payload messages - another way of saying, I don't know if what you would like to do is in the universe of possible things. Could be, I just don't know. There is quite a lot of information about analytics related to messages on the google firebase documentation site though, might have what you need already?

I don't believe the attributes you mention matter if it has a notification payload but this is subtle and I'd defer to any upstream docs from Google and/or Apple.

**Reply by jamie-unmind** [reply]
@mikehardy thank you for these clarifications!

If we were sending a data-only message to an iOS device with `contentAvailable: true`, should our background message handler be triggered even if the app has been closed & swiped away?

**Reply by mikehardy** [reply]
That case (swiped away, which is known as "force quit" most of the time I think) is one of the unreliabilities of data-only messages. iOS takes that as an explicit signal by the user to stop running the app, and acts accordingly: it will not run the app again until the user starts it again. No data-only messages will be delivered, no handlers will be called.
